hax/hax/util.py

```diff
@@ -639,33 +639,129 @@ def get_node_fid(self, node: str) -> Optional[Fid]:
+    def get_node_ctrl_fids(self, node: str) -> Optional[List[Fid]]:
         """
-        Returns the fid of the controller for the given node.
-
         Parameters:
             node : hostname of the node.
         """
         # Example,
-        # {
+        # [
         #    "key": "m0conf/sites/0x5300000000000001:0x1/
         #            racks/0x6100000000000001:0x2/encls/
         #            0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5",
-        # },
+        # ]
         encl_fid = self.get_node_encl_fid(node)
         if not encl_fid:
             return None
         ctrl_items = self.kv.kv_get('m0conf/sites', recurse=True)
         regex = re.compile(
             f'^m0conf\\/.*\\/racks\\/.*\\/encls\\/{encl_fid}\\/ctrls\\/'
             '([^/]+)$')
+        list_fids: List[Fid] = []
         for ctrl in ctrl_items:
             match_result = re.match(regex, ctrl['Key'])
             if not match_result:
                 continue
             ctrl_fid: str = match_result.group(1)
+            list_fids.append(Fid.parse(ctrl_fid))
+        return list_fids
+
+    @repeat_if_fails()
+    def get_io_service_sdevs(self, ioservice_fid: Fid) -> Optional[List[str]]:
+        if not ioservice_fid:
+            return None
+        # Example key m0conf/nodes/0x6e00000000000001:0x3/processes/
+        #   0x7200000000000001:0x15/services/0x7300000000000001:0x17/sdevs/
+        #   0x6400000000000001:0x18:{"path": "/dev/sdc", "state": "offline"}
+        sdev_items = self.kv.kv_get('m0conf/nodes', recurse=True)
+        regex = re.compile(
+            f'^m0conf\\/.*\\/processes\\/{ioservice_fid}\\/.*\\/sdevs\\/.*$')
+        sdev_fids = []
+        for sdev in sdev_items:
+            match_result = re.match(regex, sdev['Key'])
+            if not match_result:
+                continue
+            sdev_key: str = match_result.group(0)
+            key = sdev_key.split('/')
+            sdev_fids.append(key[8])
+        return sdev_fids
+
+    @repeat_if_fails()
+    def get_ioservice_ctrl_fid(self, ioservice_fid: Fid) -> Optional[Fid]:
+        """
+        Returns the fid of the controller for the given IOservice.
+
+        Parameters:
+            ioservice_fid : Fid of IO service for which ctrl fid is required.
+
+        TODO: Instead of comparing devices, add a mapping from controller
+              to I/O service
+        """
+        if not ioservice_fid:
+            return None
+
+        sdev_fids = self.get_io_service_sdevs(ioservice_fid)
+
+        if not sdev_fids:
+            return None
+
+        sdev_fid = sdev_fids[0]
+
+        # Example key m0conf/sites/0x5300000000000001:0x1/racks/
+        #    0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/
+        #    0x6300000000000001:0x5/drives/0x6b00000000000001:0x38:
+        #    {"sdev": "0x6400000000000001:0x37", "state": "M0_NC_UNKNOWN"}
+        sites_items = self.kv.kv_get('m0conf/sites', recurse=True)
```

Comment:
We are doing too many recursive calls to consul kv, mainly for `m0conf/nodes` and now for `m0conf/sites`. In PR https://github.com/Seagate/cortx-hare/pull/1737 I have added caching for such recursive calls. I think let's have a function get_all_sites() that will execute recursive call to consul kv for `m0conf/sites`. Please use that where `m0conf/sites` is referred in your PR. This will help identify the redundant calls. Later we can cache the result.

Reply:
Done